src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java

```diff
@@ -486,46 +496,69 @@ private boolean checkSynchronized(final DetailAST synchronizedAst, boolean useBr
+        final DetailAST startAST;
+        final DetailAST stopAST;
+
+        if (nextCase.getType() == TokenTypes.CASE_GROUP) {
+            startAST = currentCase.getLastChild();
+            stopAST = nextCase.getLastChild();
         }
+        // When we have last case group of switch statement then comment might
+        // be the sibling of that case, or it might be as child of that last
+        // case group AST, so we need all the comment from last case group
+        // till the end of switch case to verify if there is relief comment.
         else {
-            final int startLineNo = currentCase.getLineNo();
-            for (int i = endLineNo - 2; i > startLineNo - 1; i--) {
-                final int[] line = getLineCodePoints(i);
-                if (!CodePointUtil.isBlank(line)) {
-                    allThroughComment = matchesComment(reliefPattern, i + 1);
-                    break;
-                }
-            }
+            startAST = currentCase;
+            stopAST = currentCase.getParent().getLastChild();
         }
-        return allThroughComment;
+
+        final List<DetailAST> commentContent = getCommentContentNodes(startAST, stopAST);
+        return isReliefComment(commentContent, reliefPattern);
     }
 
     /**
-     * Does a regular expression match on the given line and checks that a
-     * possible match is within a comment.
+     * Find relief pattern in comment list.
      *
-     * @param pattern The regular expression pattern to use.
-     * @param lineNo The line number in the file.
-     * @return True if a match was found inside a comment.
+     * @param commentList list of comment to check
+     * @param reliefPattern The regular expression pattern to match the comment text
+     * @return true if relief comment is found
      */
-    // suppress deprecation until https://github.com/checkstyle/checkstyle/issues/11166
-    @SuppressWarnings("deprecation")
-    private boolean matchesComment(Pattern pattern, int lineNo) {
-        final String line = getLine(lineNo - 1);
-
-        final Matcher matcher = pattern.matcher(line);
-        return matcher.find()
-                && getFileContents().hasIntersectionWithComment(
-                        lineNo, matcher.start(), lineNo, matcher.end());
+    private static boolean isReliefComment(Iterable<DetailAST> commentList, Pattern reliefPattern) {
+        boolean result = false;
+        for (DetailAST detailAST : commentList) {
+            if (reliefPattern.matcher(detailAST.getText()).find()) {
+                result = true;
+                break;
+            }
+        }
+        return result;
     }
 
+    /**
+     * This method will return the comment content node between two nodes.
+     *
+     * @param fromNode starting node
+     * @param toNode last node to check
+     * @return list of commentContent node
+     */
+    private static List<DetailAST> getCommentContentNodes(DetailAST fromNode, DetailAST toNode) {
```

Comment:
Why do we need to traverse the AST like this at all? Can't we add `COMMENT_CONTENT` token to required types and compile a Collection of comment AST's in beginTree? Then we just filter Collection of comments based on whether they match the pattern, and we can check if there is a comment that applies to a given case group by starting/ending case group AST.

Reply:
I think it will be too much regexp execution. But we need only in switch-cases.  Switch is not very popular structure in java code, but comments are everywhere.
Performance.

Reply:
I am thinking performance, but yes this is a good call to not filter prematurely. List of all comment nodes, get sublist between case group and filter as we do now. This is more performant than repeated traversal, TreeWalker already does this work for us :)

Reply:
you will grow memory usage to keep all comments. traversing of small section of tree is very cheap, we always do this in Checks, it is not very visible, but findAllOfType finByType are same is performace as traverse. Switch-Cases are usually small in code, as they are not easy to manage by humans, so going throught 20-30 AST nodes will be quicker then extension of memory for List items to add more. 

We use Treewalker to let Check be called on main point of interest and do further local-in-Check traversing as requied, But we are not limited to use only visitToken.

Reply:
@Kevin222004 let’s keep comment collection as is for now, and come back to this idea after we resolve false positives in report.

Reply:
@nrmancuso please start reviewing false positive has been resolved

Reply:
Resolving this, I am good